staging/src/k8s.io/apimachinery/pkg/runtime/extension.go

```diff
@@ -18,10 +18,81 @@ package runtime
 
 import (
 	"bytes"
-	"encoding/json"
+	gojson "encoding/json"
 	"errors"
+	"fmt"
+	"reflect"
+
+	cbordirect "k8s.io/apimachinery/pkg/runtime/serializer/cbor/direct"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/structured-merge-diff/v4/value"
+
+	cbor "github.com/fxamacker/cbor/v2"
 )
 
+// RawExtension intentionally avoids implementing value.UnstructuredConverter for now because the
+// signature of ToUnstructured does not allow returning an error value in cases where the conversion
+// is not possible (content type is unrecognized or bytes don't match content type).
+func (re RawExtension) toUnstructured() (interface{}, error) {
+	if re.Raw != nil {
+		contentType, err := re.rawContentType()
+		if err != nil {
+			return nil, err
+		}
+
+		switch contentType {
+		default:
+			return nil, fmt.Errorf("cannot convert RawExtension with unrecognized content type %q to unstructured", contentType)
+		case ContentTypeJSON:
+			var u interface{}
+			if err := json.Unmarshal(re.Raw, &u); err != nil {
+				return nil, fmt.Errorf("failed to parse RawExtension bytes as JSON: %w", err)
+			}
+			return u, nil
+		case ContentTypeCBOR:
+			var u interface{}
+			if err := cbordirect.Unmarshal(re.Raw, &u); err != nil {
+				return nil, fmt.Errorf("failed to parse RawExtension bytes as CBOR: %w", err)
+			}
+			return u, nil
+		}
+	}
+
+	if re.Object != nil {
+		rv := reflect.ValueOf(re.Object)
+		u, err := value.TypeReflectEntryOf(rv.Type()).ToUnstructured(rv)
+		if err != nil {
+			return nil, err
+		}
+		return u, nil
+	}
+
+	return nil, nil
+}
+
+// selfDescribedCBOR is the CBOR encoding of the head of the "self-described CBOR" tag.
+var selfDescribedCBOR = []byte{0xd9, 0xd9, 0xf7}
+
+// rawContentType returns the value of the ContentType field, if not empty. Otherwise, it attempts
+// to select an appropriate content type based on the Raw bytes. If the heuristic selection fails, a
+// non-nil error is returned.
+func (re RawExtension) rawContentType() (string, error) {
+	// TODO: Allow for an explicit content type.
+
+	if !bytes.HasPrefix(re.Raw, selfDescribedCBOR) && gojson.Valid(re.Raw) {
```

Comment:
How expensive is this validity check?  If we're usually going to pay the cost to covert after this, are we doubling (or nearly doubling) the cost of conversion by adding this check?

Reply:
Maybe. If Raw is not JSON, it's likely to run into a syntax error almost immediately. In the worst case, Raw is valid JSON and this check needs to scan all of Raw. That scan would be redundant if immediately proceeded by a JSON unmarshal. With the proposed ContentType field populated, we would not make this check at all.

That said, this check does prevent `MarshalJSON` from returning invalid JSON, which it can do today and violates the contract of `json.Marshaler`.

I think I can rearrange this to avoid doing a validity check followed by an unmarshal.

Reply:
In the latest revision this uses cheaper heuristics and isn't called more than once per marshal.